File: node/src/domains/rag.js
// rag — a retrieval pipeline (chunk -> embed -> rank -> cite) over an owner-scoped document corpus. The dangerous
// property is CITATION SOUNDNESS + GROUNDING: every query hit carries a source span {doc_id, start, end} that is an
// in-bounds window into the CURRENT stored document (0 <= start <= end <= len), the hit text is exactly that window's
// code points, and an exact-chunk-text query self-matches at cosine 1.0 — so an answer is always traceable to a real
// source chunk and a citation can never be fabricated or stale. Ranking is DETERMINISTIC (score desc, ties by chunk_id
// asc — identical x3; scores are floats, NOT pinned) and a query returns at most k hits.
//
// IDENTITY + ISOLATION: both routes are POST mutations -> requireIdentity FIRST in the handler (the runtime already
// parsed the body / emitted 413/422 before the handler runs, so auth-before-validation holds, x3). No/invalid token ->
// 401. A document is USER-SCOPED two ways: the store key is the composite <owner><doc_id> (caller B can NEVER
// overwrite caller A's doc_id — the separator is a control char well_formed rejects, so it can't be forged), and the
// query scan filters on the authenticated owner FIELD (another owner's chunks are invisible). The owner is stamped
// from the token, never a body field, never surfaced in a hit.
import { isStrictInt, problem, requireIdentity, sendJSON, storePut, storeValues } from '../core/runtime.js';
import { chunkCount, chunkEnd, chunkSlice, chunkStart } from '../parts/chunking.js';
import { cosine, embed } from '../parts/embedding.js';
import { envInt } from '../parts/env_int.js';
import { isWellFormed, makeWellFormed } from '../parts/well_formed.js';
// state in store: ns "rag_documents" "<owner><doc_id>" -> {id, owner, text, chunks:[{ordinal,start,end,vector}]}.
// the composite key partitions by owner (no cross-owner clobber); the `owner` field also scopes the query scan.

const CHUNK_SIZE = envInt(process.env.RAG_CHUNK_SIZE, 400, 1);
const CHUNK_OVERLAP = envInt(process.env.RAG_CHUNK_OVERLAP, 80, 0);
const MAX_CHUNKS = envInt(process.env.RAG_MAX_CHUNKS, 1000, 1);
// fail-loud at startup: the stride size-overlap must be >= 1 (else no progress / an infinite chunk loop), x3
if (!(CHUNK_OVERLAP >= 0 && CHUNK_OVERLAP < CHUNK_SIZE)) {
  throw new Error('RAG_CHUNK_OVERLAP must satisfy 0 <= RAG_CHUNK_OVERLAP < RAG_CHUNK_SIZE');
}

export async function ragIngest(req, res, params, body) {
  const owner = await requireIdentity(req, res); // authn BEFORE validation (runtime already parsed the body), x3
  if (owner === null) return;
  if (!body || !isWellFormed(body.doc_id) || typeof body.text !== 'string' || body.text === '') {
    return problem(res, 422, 'invalid body');
  }
  // contain the text BEFORE it is chunked / embedded / stored (lone surrogate -> U+FFFD so no slice 5xxs on serialize)
  const text = makeWellFormed(body.text);
  const n = chunkCount(text, CHUNK_SIZE, CHUNK_OVERLAP);
  if (n > MAX_CHUNKS) return problem(res, 422, 'document too large'); // soft-DoS ceiling -> 422
  const chunks = [];
  for (let i = 0; i < n; i++) {
    const start = chunkStart(CHUNK_SIZE, CHUNK_OVERLAP, i);
    const end = chunkEnd(text, CHUNK_SIZE, CHUNK_OVERLAP, i);
    chunks.push({ ordinal: i, start, end, vector: embed(chunkSlice(text, start, end)) });
  }
  // owner from the token, never client-set; the composite key partitions by owner so B can't overwrite A's doc_id.
  // a blind put REPLACES the whole record (re-ingest = last-writer-wins; NOT a get-then-put RMW)
  await storePut('rag_documents', `${owner}${body.doc_id}`, { id: body.doc_id, owner, text, chunks });
  sendJSON(res, 201, { doc_id: body.doc_id, chunks: n });
}

export async function ragQuery(req, res, params, body) {
  const caller = await requireIdentity(req, res); // a POST is a mutation -> authn BEFORE validation, x3
  if (caller === null) return;
  if (!body || typeof body.query !== 'string' || body.query === '') return problem(res, 422, 'invalid body');
  let k = 3;
  if (body.k !== undefined) {
    if (!isStrictInt(body, 'k') || body.k < 1 || body.k > 50) {
      return problem(res, 422, 'k must be an integer between 1 and 50');
    }
    k = body.k;
  }
  const qv = embed(makeWellFormed(body.query)); // contain the query too — a lone-surrogate query embeds identically x3
  const hits = [];
  // unbounded-safe: ranked top-k — at most k hits (k clamped), never the corpus; the full-scan is the documented
  // embeddings-index-swap-at-scale limit. read-scope: only the caller's own docs (owner FIELD === caller).
  for (const d of await storeValues('rag_documents')) {
    if (d.owner !== caller) continue; // read-scoping: another owner's chunks are invisible (cross-corpus leak wall)
    for (const c of d.chunks) {
      hits.push({ // owner/vector/ordinal stay internal — never in the hit
        chunk_id: `${d.id}#${c.ordinal}`,
        text: chunkSlice(d.text, c.start, c.end),
        score: cosine(qv, c.vector),
        source: { doc_id: d.id, start: c.start, end: c.end },
      });
    }
  }
  hits.sort((a, b) => (b.score - a.score) || (a.chunk_id < b.chunk_id ? -1 : a.chunk_id > b.chunk_id ? 1 : 0)); // score desc, chunk_id asc
  sendJSON(res, 200, { top: hits.length > 0 ? hits[0].chunk_id : null, hits: hits.slice(0, k) }); // at most k hits, ever
}

File: go/internal/domains/rag/rag.go
// Package rag — a retrieval pipeline (chunk -> embed -> rank -> cite) over an owner-scoped document corpus. The
// dangerous property is CITATION SOUNDNESS + GROUNDING: every query hit carries a source span {doc_id, start, end}
// that is an in-bounds window into the CURRENT stored document (0 <= start <= end <= len), the hit text is exactly
// that window's code points, and an exact-chunk-text query self-matches at cosine 1.0 — so an answer is always
// traceable to a real source chunk and a citation can never be fabricated or stale. Ranking is DETERMINISTIC (score
// desc, ties by chunk_id asc — identical ×3; scores are floats, NOT pinned) and a query returns at most k hits.
//
// IDENTITY + ISOLATION: both routes are body-only POST mutations. PARSE -> AUTH -> SEMANTIC, identical ×3: the body
// is decoded FIRST (DecodeJSON drains + 413/422), THEN RequireIdentity (no/invalid token -> 401), THEN field
// validation — so an unauthenticated otherwise-422 body is 401, never a 422 that leaks shape. A document is
// USER-SCOPED two ways: the store key is the composite <owner>\x1f<doc_id> (caller B can NEVER overwrite caller A's
// doc_id — the \x1f separator is a control char well_formed rejects, so it can't be forged), and the query scan
// filters on the authenticated owner FIELD (another owner's chunks are invisible). The owner is stamped from the
// token, never a body field, never surfaced in a hit. Store names/shapes match the python/node impls.
package rag

import (
	"encoding/json"
	"net/http"
	"os"
	"sort"
	"strconv"

	"app/internal/core"
	"app/internal/parts/chunking"
	"app/internal/parts/embedding"
	"app/internal/parts/env_int"
	"app/internal/parts/well_formed"
)

type ragChunk struct {
	Ordinal int   `json:"ordinal"`
	Start   int   `json:"start"`
	End     int   `json:"end"`
	Vector  []int `json:"vector"`
}

type ragDoc struct {
	Id     string     `json:"id"`
	Owner  string     `json:"owner"` // the authenticated indexer; PRIVATE — scopes the query scan, never returned
	Text   string     `json:"text"`
	Chunks []ragChunk `json:"chunks"`
}

type ragSource struct {
	DocId string `json:"doc_id"`
	Start int    `json:"start"`
	End   int    `json:"end"`
}

type ragHit struct {
	ChunkId string    `json:"chunk_id"`
	Text    string    `json:"text"`
	Score   float64   `json:"score"`
	Source  ragSource `json:"source"`
}

var ragDocs = core.NewKV[string, ragDoc]("rag_documents")

var (
	ragChunkSize    = env_int.EnvInt(os.Getenv("RAG_CHUNK_SIZE"), 400, 1)
	ragChunkOverlap = env_int.EnvInt(os.Getenv("RAG_CHUNK_OVERLAP"), 80, 0)
	ragMaxChunks    = env_int.EnvInt(os.Getenv("RAG_MAX_CHUNKS"), 1000, 1)
)

func init() {
	// fail-loud at startup: the stride size-overlap must be >= 1 (else no progress / an infinite chunk loop), ×3
	if !(ragChunkOverlap >= 0 && ragChunkOverlap < ragChunkSize) {
		panic("RAG_CHUNK_OVERLAP must satisfy 0 <= RAG_CHUNK_OVERLAP < RAG_CHUNK_SIZE")
	}
}

func RagIngest(w http.ResponseWriter, r *http.Request) {
	// PARSE first (413/422 + drain), then AUTH (401), then SEMANTIC field validation. ×3.
	in, ok := core.DecodeJSON[struct {
		DocId *string `json:"doc_id"`
		Text  *string `json:"text"`
	}](w, r)
	if !ok {
		return
	}
	owner, ok := core.RequireIdentity(w, r) // authenticated mutation (no/invalid token -> 401)
	if !ok {
		return
	}
	if in.DocId == nil || !well_formed.IsWellFormed(*in.DocId) || in.Text == nil || *in.Text == "" {
		core.WriteProblem(w, 422, "invalid body")
		return
	}
	// contain the text BEFORE it is chunked / embedded / stored (lone surrogate -> U+FFFD; Go decode already substitutes)
	text := well_formed.MakeWellFormed(*in.Text)
	n := chunking.ChunkCount(text, ragChunkSize, ragChunkOverlap)
	if n > ragMaxChunks {
		core.WriteProblem(w, 422, "document too large") // soft-DoS ceiling -> 422
		return
	}
	chunks := make([]ragChunk, n)
	for i := 0; i < n; i++ {
		start := chunking.ChunkStart(ragChunkSize, ragChunkOverlap, i)
		end := chunking.ChunkEnd(text, ragChunkSize, ragChunkOverlap, i)
		chunks[i] = ragChunk{Ordinal: i, Start: start, End: end, Vector: embedding.Embed(chunking.ChunkSlice(text, start, end))}
	}
	// owner from the token, never client-set; the composite key partitions by owner so B can't overwrite A's doc_id.
	// a blind Set REPLACES the whole record (re-ingest = last-writer-wins; NOT a get-then-put RMW)
	ragDocs.Set(owner+"\x1f"+*in.DocId, ragDoc{Id: *in.DocId, Owner: owner, Text: text, Chunks: chunks})
	core.WriteJSON(w, 201, map[string]any{"doc_id": *in.DocId, "chunks": n})
}

func RagQuery(w http.ResponseWriter, r *http.Request) {
	raw, ok := core.DecodeJSON[json.RawMessage](w, r)
	if !ok {
		return
	}
	caller, ok := core.RequireIdentity(w, r) // AUTH (a POST is a mutation): no/invalid token -> 401, BEFORE validation
	if !ok {
		return
	}
	// SEMANTIC: k is decoded as RAW bytes then RequireIntRaw'd, so "100"/2.0/"many" are rejected HERE (after auth). Strict ×3.
	var in struct {
		Query *string         `json:"query"`
		K     json.RawMessage `json:"k"`
	}
	if json.Unmarshal(raw, &in) != nil {
		core.WriteProblem(w, 422, "invalid body")
		return
	}
	k := 3
	if in.K != nil {
		v, valid := core.RequireIntRaw(in.K) // STRICT: integer literal only (×3 with python StrictInt + node isStrictInt)
		if !valid || v < 1 || v > 50 {
			core.WriteProblem(w, 422, "k must be an integer between 1 and 50")
			return
		}
		k = v
	}
	if in.Query == nil || *in.Query == "" {
		core.WriteProblem(w, 422, "invalid body")
		return
	}
	qv := embedding.Embed(well_formed.MakeWellFormed(*in.Query)) // contain the query too — a lone-surrogate query embeds identically ×3
	hits := []ragHit{}
	// unbounded-safe: ranked top-k — at most k hits (k clamped), never the corpus; the full-scan is the documented
	// embeddings-index-swap-at-scale limit. read-scope: only the caller's own docs (owner FIELD == caller).
	for _, d := range ragDocs.All() {
		if d.Owner != caller { // read-scoping: another owner's chunks are invisible (cross-corpus leak wall)
			continue
		}
		for _, c := range d.Chunks {
			hits = append(hits, ragHit{ // owner/vector/ordinal stay internal — never in the hit
				ChunkId: d.Id + "#" + strconv.Itoa(c.Ordinal),
				Text:    chunking.ChunkSlice(d.Text, c.Start, c.End),
				Score:   embedding.Cosine(qv, c.Vector),
				Source:  ragSource{DocId: d.Id, Start: c.Start, End: c.End},
			})
		}
	}
	sort.Slice(hits, func(i, j int) bool { // DETERMINISTIC: score desc, ties by chunk_id asc
		if hits[i].Score != hits[j].Score {
			return hits[i].Score > hits[j].Score
		}
		return hits[i].ChunkId < hits[j].ChunkId
	})
	if len(hits) > k {
		hits = hits[:k] // at most k hits, ever (clamp — no panic when k > len)
	}
	var top any
	if len(hits) > 0 {
		top = hits[0].ChunkId
	}
	core.WriteJSON(w, 200, map[string]any{"top": top, "hits": hits})
}
